### PR TITLE
GH-89769: Pathlib - do not follow links when checking for precise glob match

### PR DESCRIPTION
**Pull request by akulakov**
https://bugs.python.org/issue45606

I've also added a *follow_symlinks* arg to `exists()` method because it seems more logical to be able to test if a path exists via the same method rather than having to also remember and to check for it being a symlink. This is also consistent with a few other *Path* methods; and it makes it more explicit and clearer that by default `exists()` follows symlinks, by looking at its signature (otherwise this may be a surprising behaviour).

* Issue: gh-89769

### CODE REVIEW

File: Lib/test/test_pathlib.py [hunk 1]
```diff
@@ -1529,6 +1529,8 @@ def test_exists(self):
         self.assertIs(False, P('/xyzzy').exists())
         self.assertIs(False, P(BASE + '\udfff').exists())
         self.assertIs(False, P(BASE + '\x00').exists())
+        self.assertIs(False, (p / 'brokenLink').exists())
+        self.assertIs(True, (p / 'brokenLink').exists(follow_symlinks=False))
```
Comment (ambv) [inline Lib/test/test_pathlib.py]: Shouldn't this also use `if os_helper.can_symlink():`?

**Review by MaxwellDupre — APPROVED** [review]
Works on my Ubuntu and I see bot tests are ok, so ok with me.

File: Doc/library/pathlib.rst [hunk 2]
```diff
@@ -755,9 +755,14 @@ call fails (for example because the path doesn't exist).
    .. versionchanged:: 3.10
       The *follow_symlinks* parameter was added.
 
-.. method:: Path.exists()
+.. method:: Path.exists(follow_symlinks=True)
```
Comment (barneygale) [inline Doc/library/pathlib.rst]:
```suggestion
.. method:: Path.exists(*, follow_symlinks=True)
```

File: Lib/pathlib.py [hunk 3]
```diff
@@ -422,7 +422,8 @@ def __init__(self, name, child_parts, flavour):
     def _select_from(self, parent_path, is_dir, exists, scandir):
         try:
             path = parent_path._make_child_relpath(self.name)
-            if (is_dir if self.dironly else exists)(path):
+            if ((self.dironly and is_dir(path)) or \
+                    (not self.dironly and exists(path, follow_symlinks=False))):
```
Comment (barneygale) [inline Lib/pathlib.py]:
```suggestion
            follow = is_dir(path) if self.dironly else exists(path, follow_symlinks=False)
            if follow:
```

File: Lib/pathlib.py [hunk 4]
```diff
@@ -1280,12 +1281,16 @@ def link_to(self, target):
 
     # Convenience functions for querying the stat results
 
-    def exists(self):
+    def exists(self, follow_symlinks=True):
```
Comment (barneygale) [inline Lib/pathlib.py]:
```suggestion
    def exists(self, *, follow_symlinks=True):
```

File: Lib/pathlib.py [hunk 5]
```diff
@@ -1280,12 +1281,16 @@ def link_to(self, target):
...
         """
         Whether this path exists.
+
+        If path is a symlink: with `follow_symlinks=True`, return True if path
+        pointed to by the symlink exists; with follow_symlinks=False, return
+        True if the symlink itself exists.
```
Comment (barneygale) [inline Lib/pathlib.py]: This seems unnecessary detail to me for a docstring. It's fairly self-explanatory I think.

File: Misc/NEWS.d/next/Library/2021-11-19-23-37-18.bpo-45606.UW5XE1.rst [hunk 6]
```diff
@@ -0,0 +1,5 @@
+Fixed the bug in :meth:`~pathlib.Path.glob` -- previously a dangling symlink
```
Comment (barneygale) [inline Misc/NEWS.d/next/Library/2021-11-19-23-37-18.bpo-45606.UW5XE1.rst]:
```suggestion
Fixed the bug in :meth:`pathlib.Path.glob` -- previously a dangling symlink
```

**Review by barneygale — CHANGES_REQUESTED** [review]
Change looks good! Could you fix conflicts please?

File: Lib/pathlib.py (hunk 5 above)
Comment (akulakov) [reply Lib/pathlib.py]: I was a bit concerned about users who are not familiar with dangling symlinks, in that case this param would be confusing without explanation. I think I was either unfamiliar with them or haven't seen them in too long of a time when I was working on this issue. If you feel most users of this method should be familiar with them or be able to look it up, I will remove the note, just wanted to point out that was my reason for adding it.

Comment (barneygale) [reply Lib/pathlib.py]: Makes sense. Maybe a small wording tweak, then?

> This method normally follows symlinks; to check whether a symlink exists, add the argument ``follow_symlinks=False``.

File: Doc/library/pathlib.rst [hunk 7]
```diff
@@ -768,10 +773,8 @@ call fails (for example because the path doesn't exist).
       >>> Path('nonexistentfile').exists()
       False
 
-   .. note::
-      If the path points to a symlink, :meth:`exists` returns whether the
-      symlink *points to* an existing file or directory.
-
+   .. versionchanged:: 3.11
```
Comment (barneygale) [inline Doc/library/pathlib.rst]:
```suggestion
   .. versionchanged:: 3.12
```

**Review by barneygale — APPROVED** [review]
Last nit. Thanks!

**Comment by barneygale** [comment]
Thanks again, @akulakov

**Comment by akulakov** [comment]
Thanks @barneygale  for the review and merging!

**Pull request merged**